File: libs/nats/pkg/tools.go
package pkg

import (
	"encoding/json"
	"github.com/dehwyy/mafuyu/libs/nats/pkg/errors"
	"github.com/nats-io/nats.go/jetstream"
	"strings"
)

type Tools struct {
	Msg jetstream.Msg
}

func NewTools(msg jetstream.Msg) *Tools {
	return &Tools{Msg: msg}
}

func (t *Tools) GetSubject() (string, error) {
	parts := strings.SplitN(t.Msg.Subject(), ".", 3)
	if len(parts) < 3 {
		return "", errors.NewNatsError("malformed subject")
	}

	v := parts[2]
	return v, nil
}

func GetPayload[T any](b []byte) (T, error) {
	var v T
	if err := json.Unmarshal(b, &v); err != nil {
		return v, err
	}

	return v, nil
}

File: libs/nats/src/tools.rs
use super::errors::{RouteError, NatsHandleError};

pub struct Tools;

impl Tools {
    /// # Subject Convention.
    /// - Starts with `tokens`.
    /// - Second path's part either `call` or `do`.
    /// - At least `3` path's parts
    /// ## Description.
    /// - `call` means `request-reply` scenario
    /// - `do` means `act as jetstream` (async communication, w/o response).
    /// Example of valid `path` is `tokens.call.validate_token` or `tokens.do.clear.all`
    pub fn get_subject(initial_subject: &async_nats::Subject) -> String {
        // Let's simplify :)
        initial_subject.to_string()

        // let parts = initial_subject.splitn(3, ".").collect::<Vec<&str>>();

        // let parsed =parts.get(2).map(|v| v.to_string());

        // match parsed {
        //     Some(s) => Ok(s),
        //     None => Err(SubjectError::MalformedSubjectName("[invalid subject path]: @see(src/internal/tools.rs)".to_string()))
        // }
    }

    pub fn get_payload<'a, T>(b: &'a [u8]) -> Result<T, RouteError>
    where T: serde::Deserialize<'a> {
        crate::message::Decoder::decode(b).invalid_argument_error()
    }
}